conans/client/manager.py

```diff
@@ -92,16 +91,27 @@ def deps_install(app, ref_or_path, install_folder, graph_info, remotes=None, bui
                                      interactive=manifest_interactive)
         manifest_manager.print_log()
 
+    conanfile.folders.set_base_install(install_folder)
+    conanfile.folders.set_base_imports(install_folder)
+    conanfile.folders.set_base_generators(base_folder)
+
+    output = conanfile.output if root_node.recipe != RECIPE_VIRTUAL else out
+
     if install_folder:
-        conanfile.layout.set_base_install_folder(install_folder)
         # Write generators
-        output = conanfile.output if root_node.recipe != RECIPE_VIRTUAL else out
-        if generators is not False:
-            tmp = list(conanfile.generators)  # Add the command line specified generators
-            tmp.extend([g for g in generators if g not in tmp])
-            conanfile.generators = tmp
-            app.generator_manager.write_generators(conanfile, install_folder, output)
-            write_toolchain(conanfile, install_folder, output)
+        tmp = list(conanfile.generators)  # Add the command line specified generators
+        generators = set(generators) if generators else set()
+        tmp.extend([g for g in generators if g not in tmp])
+        conanfile.generators = tmp
+        app.generator_manager.write_generators(conanfile, install_folder,
+                                               conanfile.generators_folder,
+                                               output)
+        write_toolchain(conanfile, conanfile.generators_folder, output)
+        if add_txt_generator:
```

Comment:
Why not add ``txt`` before and let the first call to ``write_generators()`` hadle it?